content/blog/next-level-iac-briding-the-declarative-gap/index.md

```diff
@@ -0,0 +1,164 @@
+
+Using the `apply` function we can pass an `Input<T>` or `Output<T>` to a [transformation function][data-transformation-wiki] which will accept the realized value as its input (and only run when that value is available). Within that function you can handle the string as a normal string, then return it back as an `Output<string>`. This allows you to use the full power of your programming language to perform a custom operation inline with the rest of your infrastructure automation.
+
+Here's a simple example of using `apply` to transform an output to something new using your programming language's libraries. In this case, [Base64 encoding][base64-wiki] the `id` of our `bucket` object.
+
+### Example: Using `apply` to Base64 encode an Output<string>
+
+{{< chooser language "typescript,python" />}}
+
+{{% choosable language typescript %}}
+
+```typescript {hl_lines=["10-14"]}
+import * as pulumi from '@pulumi/pulumi';
+import * as aws from '@pulumi/aws';
+
+// Create an S3 bucket
+const bucket = new aws.s3.Bucket("bucket");
+
+// Export the name of the bucket
+export const bucketName = bucket.id;
+
+// Export the name of the bucket, in Base64 encoding
+export const bucketNameBase64 = bucket.id.apply(id=> {
+    // Transform the bucket name into a Base64 encoded string
+    Buffer.from(id).toString('base64')
+})
+```
+
+{{% /choosable %}}
+
+{{% choosable language python %}}
+
+```python {hl_lines=[1, "11-19"]}
+import base64
+import pulumi
+import pulumi_aws as aws
+
+# Create an S3 bucket
+bucket = aws.s3.Bucket('bucket')
+
+# Export the name of the bucket
+pulumi.export('bucket_name', bucket.id)
+
+# A Base64 encoder function to use in the apply function's lambda
+def base64_encode(data):
+    data_bytes = data.encode("ascii")
+    base64_bytes = base64.b64encode(data_bytes)
+    base64_string = base64_bytes.decode("ascii")
+
+# Export the name of the bucket, in Base64 encoding
+pulumi.export('bucket_name', bucket.id.apply(lambda data: base64_encode))
+```
+
+{{% /choosable %}}
+
+Our example shows how to do Base64 encoding, but this technique can be used for any kind of [transformation function][data-transformation-wiki]. These can be as elaborate as your use case requires, and leverage the full power of your [programming language and its entire ecosystem][next-level-iac-package-ecosystems]. Doing custom transformations like this in other tools, when possible, requires using external coordination which breaks the declarative model and interrupts the deployment engine's ability to fully understand your infrastructure's dependency graph.
```

Comment:
I feel like it would be good to compare and contrast Pulumi with a system that doesn't have promises. What does it look like. What is not able to be done. Like is this completely not doable in TF or do they have a workaround here that you get like sorta that functionality. I believe in CFN its completely undoable because it just executes in order of the template

Reply:
Yeah, It's tempting to compare/contrast, but sticking with the idea that our audience is already familiar with TF and other tools, they should already understand the pain of doing these kinds of operations.